### PR TITLE
Allow multiple chromedriver versions to be available

### PR DESCRIPTION
**Pull request by imurchie**
Allow the user to have multiple chromedriver executables available, and pick the most recent one that supports the Chrome being automated. Pass in `executableDir` (defaulting to the directory installed with this package) where the executables are found, so user can have a permanent store of chromedriver, should they wish.

### CODE REVIEW

File: lib/chromedriver.js [hunk 1]
```diff
@@ -17,29 +20,89 @@ const DEFAULT_PORT = 9515;
...
     this.verbose = verbose;
     this.logPath = logPath;
   }
 
+  async getCompatibleChromedriver () {
+    if (!this.adb) {
+      return await getChromedriverBinaryPath();
+    }
+
+    const chromeVersion = (semver.coerce(await getChromeVersion(this.adb, this.bundleId)));
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: extra parentheses here

Comment (mykola-mokhnach) [reply lib/chromedriver.js]: do we need to verify if chromeVersion is coerced properly?

File: lib/chromedriver.js [hunk 2]
```diff
@@ -17,29 +20,89 @@ const DEFAULT_PORT = 9515;
...
+    const mapping = JSON.parse(await fs.readFile(path.resolve(__dirname, '..', '..', 'versions.json')));
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: does JSON.parse work with buffers?

File: lib/chromedriver.js [hunk 3]
```diff
@@ -17,29 +20,89 @@ const DEFAULT_PORT = 9515;
...
+
+    // go through the versions available
+    const executables = await fs.glob(`${this.executableDir}/*`);
+    const cds = (await asyncmap(executables, async function (executable) {
+      try {
+        const {stdout} = await exec(executable, ['--version']);
+        const match = /ChromeDriver\s(\d+\.\d+\.\d+)\s/.exec(stdout);
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: shall we also use semver here?

File: lib/chromedriver.js [hunk 4]
```diff
@@ -17,29 +20,89 @@ const DEFAULT_PORT = 9515;
...
+    })).filter((cd) => !!cd);
+    log.debug(`The following Chromedriver executables were found:`);
+    for (const cd of cds) {
+      log.debug(`    ${cd.executable} (minimum Chrome version '${cd.minCDVersion}')`);
+    }
+
+    const workingCds = cds.filter((cd) => semver.gte(chromeVersion, cd.minCDVersion))
+      .sort((a, b) => a.version <= b.version ? 1 : -1);
+
+    if (!workingCds.length) {
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: _.isEmpty is more readable IMHO

File: lib/utils.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+async function getChromeVersion (adb, bundleId) {
+  const output = await adb.getPackageInfo(bundleId);
```
Comment (mykola-mokhnach) [inline lib/utils.js]: `const {versionName} =`

File: lib/chromedriver.js [hunk 6]
```diff
@@ -17,29 +20,89 @@ const DEFAULT_PORT = 9515;
...
+          };
+        }
+      } catch (ign) {}
+    })).filter((cd) => !!cd);
+    log.debug(`The following Chromedriver executables were found:`);
+    for (const cd of cds) {
+      log.debug(`    ${cd.executable} (minimum Chrome version '${cd.minCDVersion}')`);
+    }
+
+    const workingCds = cds.filter((cd) => semver.gte(chromeVersion, cd.minCDVersion))
```
Comment (mykola-mokhnach) [inline lib/chromedriver.js]: do we start from the lowest or from the highest compatible version here?

**Comment by mykola-mokhnach** [comment]
It would be nice to have at least one unit tests, which verifies versions matching algorithm.

**Comment by mykola-mokhnach** [comment]
so testing, much unit, WOW

File: lib/chromedriver.js [hunk 7]
```diff
@@ -4,42 +4,145 @@ import events from 'events';
...
+import { getChromeVersion } from './utils';
+import semver from 'semver';
+import _ from 'lodash';
 
 
 const log = logger.getLogger('Chromedriver');
 
 const DEFAULT_HOST = '127.0.0.1';
 const DEFAULT_PORT = 9515;
+const CHROMEDRIVER_CHROME_MAPPING = {
```
Comment (aluedeke) [inline lib/chromedriver.js]: this doesn't really solve the problem of Appium not supporting versions of chrome which got released after the release of Appium. does it? Can we put that into a json which may get downloaded from git?

Comment (aluedeke) [inline lib/chromedriver.js]: we are also looking for a solution which we can use to batch download the drivers and bundle it into a docker image

Comment (imurchie) [reply lib/chromedriver.js]: No, and it's not intended to. That can be handled with `chromedriverExecutable`.

Comment (imurchie) [reply lib/chromedriver.js]: That can be done if you download them somewhere and then pass in the executable bin, which will eventually be a capability for the android/uiautomator2 drivers.

**Pull request merged**